caddytls/ctmonitor.go

```diff
@@ -0,0 +1,244 @@
+				for i, issuance := range issuanceObjects {
+					bytes := decodeField(issuance.Cert.Data)
+					aKey := string(bytes)
+					if _, ok := retrievedCerts[aKey]; ok {
+						continue
+					} else {
+						if issuance.Cert.Type == "precert" {
+							continue
+						}
+						value := "ID: " + issuance.Cert.ID + " " + issuance.Issuer.Name + " not valid before: " + issuance.NotBefore +
+							" and not valid after: " + issuance.NotAfter
+						retrievedCerts[aKey] = value
+					}
+					if i == (len(issuanceObjects) - 1) {
+						issuanceId, err := strconv.Atoi(issuance.ID)
+						if err != nil {
+							log.Printf(err.Error())
+							log.Print("Error occured on line 162 of ctmonitor")
+						}
+						if issuanceId > biggestId {
+							biggestId = issuanceId
+						}
+						concurrent = issuanceId
+					}
+				}
+			} else {
+				finalResponse, err := http.Get(query + prepQuery(domainName,
+					subdomains, wildcards, strconv.Itoa(concurrent)))
+				if err != nil {
+					log.Fatalf("https get request failed on input %s \nError: %v", prepQuery(domainName, subdomains, wildcards, strconv.Itoa(concurrent)), err)
+				}
+				defer finalResponse.Body.Close()
+				retryAfter = finalResponse.Header.Get("Retry-After")	
+			}
+		}
+	}
+	putLatestId(biggestId, FILE_PATH)
+	timeToWait, err := strconv.Atoi(retryAfter)
+	if err != nil {
+		log.Printf(err.Error())
+		log.Print("Error occured on line 185 of ctmonitor, returning 1 hour")
+		timeToWait = 3600
+	}
+	//check(err)
+	return retrievedCerts, timeToWait
+}
+
+// monitorCerts continuously monitors the certificates that Caddy serves. monitorCerts queries again 
+// after the specified time.
+func monitorCerts() {
+	//var fetchedCerts map[string]string
+	for {
+		namesToLookUp, caddyCerts := getCaddyCerts()
+		startingIndex, err := getLatestIndex(FILE_PATH)
+		if err != nil {
+			log.Printf("Error %v while getting starting index, starting at 0", err.Error())
+		}
+		fetchedCerts, pause := lookUpNames(namesToLookUp, BASE_URI, false, false, startingIndex)
+		compareCerts(caddyCerts, fetchedCerts)
+		time.Sleep(time.Duration(pause) * time.Second)
+	}
+}
+
+func prepQuery(domainName string, subdomains bool, wildcards bool, index string) string {
+	v:= url.Values{}
+	v.Set("domain", domainName)
+	if wildcards {
+		v.Set("match_wildcards", "true")
+	}
+	if subdomains {
+		v.Set("include_subdomains", "true")
+	}
+	v.Set("after", index)
+	v.Add("expand", "dns_names")
+	v.Add("expand", "issuer")
+	v.Add("expand", "cert")
+	encodedValue := v.Encode()
+	
+	return encodedValue
+}
+
+func putLatestId(currentId int, fileName string) {
+	fileHandle, err := os.Create(fileName)
+	if err != nil {
+		log.Printf("[ERROR] could not write to %v", fileName)
+	}
+	writer := bufio.NewWriter(fileHandle)
+	defer fileHandle.Close()
+	writeValue := strconv.Itoa(currentId)
+	fmt.Fprintln(writer, writeValue)
+	writer.Flush()
+}
+
+func startMonitoring(eventType caddy.EventName, eventInfo interface{}) error {
+	go monitorCerts()
```

Comment:
Check the event type to make sure this is the right event -- but let's look at this in more detail after your first pass

Reply:
How would I go about testing this portion?  Is it possible to write a test that would send the startupEvent signal and then stop it with the ShutdownEvent?